lib/util/version.js

```diff
@@ -21,14 +22,36 @@ function resetDetectedVersion() {
   cachedDetectedReactVersion = undefined;
 }
 
+function resolveBasedir(context) {
+  let basedir;
+  if (context) {
+    const filename = context.getFilename();
+    const dirname = path.dirname(filename);
+    try {
+      if (fs.statSync(filename).isFile()) {
+        // dirname must be dir here
+        basedir = dirname;
+      }
+    } catch (err) {
+      // https://github.com/eslint/eslint/issues/11989
+      if (err.code === 'ENOTDIR') {
+        // the error code alreay indicates that dirname is a file
+        basedir = path.dirname(dirname);
+      }
+    }
+  }
+  // fallback to cwd only if basedir not resolved
+  return basedir || process.cwd();
+}
+
 // TODO, semver-major: remove context fallback
 function detectReactVersion(context) {
   if (cachedDetectedReactVersion) {
     return cachedDetectedReactVersion;
   }
 
   try {
-    const reactPath = resolve.sync('react', {basedir: context ? path.dirname(context.getFilename()) : process.cwd()});
+    const reactPath = resolve.sync('react', {basedir: resolveBasedir(context)});
```

Comment:
can `resolveBasedir` be moved up out of this try block?

Reply:
I don't see any difference. The caught error will be rethrown.

Reply:
Code that never throws shouldn't be inside a `try`, or it will run more slowly for no reason.

Reply:
Is there any reference that `try` is `slow`? I'd prefer cleaner codes.

Reply:
Either way, it's far cleaner to only put "code that can throw" inside of `try` than to avoid calling `process.cwd()`.